code/datums/position_point_vector.dm

```diff
@@ -0,0 +1,258 @@
+	AM.forceMove(return_turf())
+	AM.pixel_x = return_px()
+	AM.pixel_y = return_py()
+
+/datum/point/proc/return_turf()
+	return locate(Ceiling(x / world.icon_size, 1), Ceiling(y / world.icon_size, 1), z)
+
+/datum/point/proc/return_coordinates()		//[turf_x, turf_y, z]
+	return list(Ceiling(x / world.icon_size, 1), Ceiling(y / world.icon_size, 1), z)
+
+/datum/point/proc/return_position()
+	return new /datum/position(src)
+
+/datum/point/proc/return_px()
+	return Modulus(x, world.icon_size) - 16 - 1
+
+/datum/point/proc/return_py()
+	return Modulus(y, world.icon_size) - 16 - 1
+
+/datum/point/proc/mapcheck()
+	return qdel(src)			//Random maps = don't bother.
+	/*
+	. = FALSE
+	var/maxx = world.icon_size * world.maxx
+	var/maxy = world.icon_size * world.maxy
+	var/move_zx = 0
+	var/move_zy = 0
+	if(x < 0)
+		x += maxx
+		move_zx -= 1
+	if(y < 0)
+		y += maxy
+		move_zy -= 1
+	if(x > maxx)
+		x -= maxx
+		move_zx += 1
+	if(y > maxy)
+		y -= maxy
+		move_zy += 1
+	var/datum/space_level/S = SSmapping.get_level(z)
+	if(move_zx != 0)
+		var/datum/space_level/L = S.neigbours["[move_zx < 0? WEST : EAST]"]
+		z = L.z_value
+		. = TRUE
+	if(move_zy != 0)
+		var/datum/space_level/L = S.neigbours["[move_zy < 0? SOUTH : NORTH]"]
+		z = L.z_value
+		. = TRUE
+	*/
+
+/datum/point/vector
+	var/speed = 32				//pixels per iteration
+	var/iteration = 0
+	var/angle = 0
+	var/mpx = 0					//calculated x/y movement amounts to prevent having to do trig every step.
+	var/mpy = 0
+	var/starting_x = 0			//just like before, pixels from EDGE of map! This is set in initialize_location().
+	var/starting_y = 0
+	var/starting_z = 0
+
+/datum/point/vector/New(_x, _y, _z, _pixel_x = 0, _pixel_y = 0, _angle, _speed, initial_increment = 0)
+	..()
+	initialize_trajectory(_speed, _angle)
+	if(initial_increment)
+		increment(initial_increment)
+
+/datum/point/vector/initialize_location(tile_x, tile_y, tile_z, p_x = 0, p_y = 0)
+	. = ..()
+	starting_x = x
+	starting_y = y
+	starting_z = z
+
+/datum/point/vector/copy_to(datum/point/vector/v = new)
+	..(v)
+	v.speed = speed
+	v.iteration = iteration
+	v.angle = angle
+	v.mpx = mpx
+	v.mpy = mpy
+	v.starting_x = starting_x
+	v.starting_y = starting_y
+	v.starting_z = starting_z
+	return v
+
+/datum/point/vector/proc/initialize_trajectory(pixel_speed, new_angle)
+	if(!isnull(pixel_speed))
+		speed = pixel_speed
+	set_angle(new_angle)
+
+/datum/point/vector/proc/set_angle(new_angle)		//calculations use "byond angle" where north is 0 instead of 90, and south is 180 instead of 270.
+	if(isnull(angle))
+		return
+	angle = new_angle
+	update_offsets()
+
+/datum/point/vector/proc/update_offsets()
+	mpx = sin(angle) * speed
+	mpy = cos(angle) * speed
+
+/datum/point/vector/proc/set_speed(new_speed)
+	if(isnull(new_speed) || speed == new_speed)
+		return
+	speed = new_speed
+	update_offsets()
+
+/datum/point/vector/proc/increment(multiplier = 1)
+	iteration++
+	x += mpx * 1
+	y += mpy * 1
+	if(mapcheck())
```

Comment:
This will delete vector after increment always. Is this intended?

Reply:
no

Reply:
fixed